[next] Update docs (#92)
toomuchdesign/next-page-tester: README.md

The missing DOM integration testing tool for [Next.js][next-github].

> Given a Next.js route, this library will **mount the matching page in JSDOM**, provided with the expected **props** derived from Next.js' [**routing system**][next-docs-routing] and [**data fetching methods**][next-docs-data-fetching].

Reviewer: ```suggestion
Given a Next.js route, this library will **render the matching page in JSDOM**, provided with the expected **props** derived from Next.js' [**routing system**][next-docs-routing] and [**data fetching methods**][next-docs-data-fetching].
```

Given a Next.js route, this library will **mount the matching page in JSDOM**, provided with the expected **props** derived from Next.js' [**routing system**][next-docs-routing] and [**data fetching methods**][next-docs-data-fetching].

```js

> import { render, screen, fireEvent } from '@testing-library/react';

Reviewer: Would it make sense to make this a section with "Using @testing-library/react" header and provide another example how library could be used without it?

Author: What if we did the opposite? :) Like adding a section where we explain that `next-page-tester` does nothing more then appending stuff in in the (mocked) DOM and call React.hydrate on it.

I assume `@testing-library` is almost a de-facto standard library and I'd like to not clutter the initial example with too 
many DOM API's (considering we simulate a click event). How do you see it?

Reviewer: Perhaps we should create an `examples` folder where we can "gather" some patterns to not clutter `README.md`. It can be an ongoing effort as a "response" to potential issues users might face.

## What

> The idea behind this library is to reproduce as closer as possible the way Next.js works without spinning up servers, and render the output in a local JSDOM environment.

Reviewer: ```suggestion
The idea behind this library is to reproduce as closely as possible the way Next.js works without spinning up servers, and render the output in a local JSDOM environment.
```

The mounted application is **interactive** and can be tested with any DOM testing library (like [`@testing-library/dom`](https://testing-library.com/docs/dom-testing-library/intro)).

`next-page-tester` will take care of:

- load and execute modules in the **expected browser or server environments**
- **resolving** provided **routes** into matching page component
- calling **Next.js data fetching methods** (`getServerSideProps`, `getInitialProps` or `getStaticProps`) if the case
- set up a **mocked `next/router` provider** initialized with the expected values (to test `useRouter` and `withRouter`)
- wrapping page with custom `_app` and `_document` components
- emulating **client side navigation** via `Link`, `router.push`, `router.replace`
- handling pages' `redirect` returns

## API

### getPage

`getPage` accepts an [option object](/#options) and returns 4 values:

```ts
function getPage(options: Options): Promise<
  renderHTML: () => HTMLElement<NextRoot>
  render: () => HTMLElement<NextRoot>
  html: string
  page: React.ReactElement<AppElement>
>;
```

#### `renderHTML()`

Inject server side rendering output to the DOM. Returns `#__next` root element element.

#### `render()`

> Calls `renderHTML()` and mounts/hydrate the the application on `#__next` root element. Returns `#__next` root element element.

Reviewer: Lets clarify when user would want to use each option. If I understand it correctly, `render` should be used in most cases, but `renderHTML` could be used if users want more control and do the `hydration` themselves?

Author: Yep `hydration` themselves but especially test what a search engine would see. 👍

#### `html`

> HTML string representing server side rendering output.

Reviewer: ```suggestion
HTML string representing output of server side rendering.
```

#### `page`

> React element the application

Reviewer: ```suggestion
React element of the application
```

3. **mount/hydrate** the client application to the previously rendered html

The mounted application is **interactive** and can be tested with any DOM testing library (like [`@testing-library/dom`](https://testing-library.com/docs/dom-testing-library/intro)).

`next-page-tester` will take care of:

- load and execute modules in the **expected browser or server environments**
- **resolving** provided **routes** into matching page component
- calling **Next.js data fetching methods** (`getServerSideProps`, `getInitialProps` or `getStaticProps`) if the case
- set up a **mocked `next/router` provider** initialized with the expected values (to test `useRouter` and `withRouter`)
- wrapping page with custom `_app` and `_document` components
- emulating **client side navigation** via `Link`, `router.push`, `router.replace`
- handling pages' `redirect` returns

## API

### getPage

`getPage` accepts an [option object](#options) and returns 4 values:

```js
import { getPage } from 'next-page-tester';
const { render, renderHTML, html, page } = await getPage({ options });
```

### render()

Type: `() => HTMLElement<NextRoot>`<br/>
Returns: `#__next` root element element.

> You will mostly use this render method. It calls `renderHTML()` and then mounts/hydrates the React application into JSDOM `#__next` root element. This is what the user gets when she visits a page.

Reviewer: ```suggestion
Unless you have an advances use-case, you should mostly use this method. Under the hood it calls a lower level `renderHTML()` method and then mounts/hydrates the React application into JSDOM `#__next` root element. This is what the user would get/see when they visit a page.
```

> ### renderHTML()

Reviewer: What do you think about this?

```suggestion
### prepareDocument()
```

Author: I'll try to put down what I think `renderHTML` does compared to a real Next.js app.

From my point of `renderHTML()/whatever` represents the usual initial rendering process which happens in any browser when a user visits any HTML page: HTML file download, parsed and a DOM tree generated out if it.

For a static HTML page the game ends with `renderHTML()/whatever`.

The exact same thing happens for a SSRendered React app but there's is an extra step: when JS resources are loaded React takes over, and does whatever it was told to do with the just rendered DOM.

From users' perspective, I'd say that `renderHTML()` renders the `html` string in the DOM (or any equivalent word for transforming a string into it's visual representation)

`prepareDocument` sounds a bit like "setting up the environment before we do anything", while `renderHTML()` paints the whole page markup on the screen (except for the detail that JSDOM doesn't provide visual output 😄 ).

`renderHTML()` should be something close to:
- `renderSSRResultInDOM`
- `renderSSR`
- `renderStaticHTML`
- `renderSSRHTML`

`render()` should be something like:
- `renderAndMountPage`
- `renderPageAndMountReact`
- `renderSSRAndMountPage`
- `renderAndHydrate`

Reviewer: `renderStaticMarkup` & `render`?

I think `renderStaticMarkup` makes it clear that output is SSR and that it's not interactive. As for the "main" `render` method I would not complicate and make name as simple as possible: `render`.

Author: I'd slim down the list to:
- renderStaticMarkup
- renderStaticDocument
- renderDocument

Taken from React docs:
> Render a React element to its initial HTML. React will return an HTML string. You can use this method to generate HTML on the server and send the markup down on the initial request for faster page loads and to allow search engines to crawl your pages for SEO purposes.

> If you call ReactDOM.hydrate() on a node that already has this server-rendered markup, React will preserve it and only attach event handlers, allowing you to have a very performant first-load experience.

- renderInitialHTML
- renderInitialMarkup
- renderServerRenderedMarkup
- renderServerRenderedHTML

(even though our server-rendered is emulated)

I'd try not to add another return method, right now. I don't see much value right now. Actually I considered removing both `html` and `page`: they have sense only if someone tried to render and hydrate on their side.
